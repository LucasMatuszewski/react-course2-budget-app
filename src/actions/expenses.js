import uuid from 'uuid';

// we make here NAMED EXPORTS (not default)

/***********
 * ACTIONS (action generators - functions to generate an Action)
 */

// ADD_EXPENSE ( property = default value )
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(), // we use uuid - NPM library to generate RFC4122 UUIDs (we have to install and import it)
        description,
        note,
        amount,
        createdAt
    }
});

// REMOVE_EXPENSE
// const removeExpense = (id) => ({ // return object: {id: "UUID"}
export const removeExpense = ({ id } = {}) => ({ // return only UUID
    type: 'REMOVE_EXPENSE',
    id
    // expense: { id } // we don't need "expense: { }" here. We only want to pass id to delete it
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});