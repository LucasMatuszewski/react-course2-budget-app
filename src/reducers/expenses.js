/***********
 * REDUCERS
 */

// Expanses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => { //default state is empty array
    switch (action.type) {
        case 'ADD_EXPENSE':
            // state.push(action.expense); // We can't use PUSH = it changes array. Reducer wold not be a PURE Function
            // state.concat(action.expense); // CONCAT() don't change array, it returns new one.
            // BUT WE CAN USE ES6 SPREAD OPERATOR:
            return [ //with spread operator we have a control where to add new item (it could be on beginning)
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // console.log(action.id);
            // console.log(state);
            // function removeId(expense) {
            //     console.log(expense.id);
            //     return expense.id != action.id
            // }
            // return state.filter((expense) => expense.id != action.id );
            return state.filter(({ id }) => id != action.id); // destructuring id from expense object
        // filter doesn't change array, returns new one
        case 'EDIT_EXPENSE':
            /* //my idea to do it. Works, but .map() is better :)
            console.log('state:', state.filter(({ id }) => id == action.id)[0]);
            return {
                ...state.filter(({id}) => id == action.id )[0],
                amount: action.updates
            }; */
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense, //spread previous values
                        ...action.updates //spread with new values, to edit / overwrite old values
                    }
                } else {
                    return expense;
                }
            });
        // INSTRUCTOR DIDN'T WANT US TO ADD THIS IN EXPENSES REDUCER... ONLY TO FILTERS REDUCER
        /* case 'SET_TEXT_FILTER':
            return state.filter(({ id, description, note, amount }) => {
                return id == action.text || description == action.text || note == action.text || amount == action.text;
                // return description == action.text;
            }); */
        default:
            return state;
    }
};

// export default expensesReducer; // we don't have to name this reducer if we export it by default