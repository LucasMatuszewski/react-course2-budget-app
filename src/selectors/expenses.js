import moment from 'moment';
/******
 * SELECTOR
 * GET FILTERED/Visable EXPENSES & Sort them
 */
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        /* OLD VERSION WITHOUT MOMENT():
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof startDate !== 'number' || expense.createdAt <= endDate; */
        // QUERY with MOMENT() https://momentjs.com/docs/#/query/is-same-or-before/
        // IMPORTANT!!! There is some error/bug because of Redux DevTools on data changes:
        // startDate.isSameOrBefore is not a function
        // After deactivation of Redux DevTools it works fine.
        // similar isue: https://github.com/zalmoxisus/redux-devtools-extension/issues/315
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; // startDate is already a moment() object, we don't have to convert it
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        // const textMatch = typeof text !== 'string' || expense.description.toLowerCase().indexOf(text.toLowerCase()) !== -1 || expense.note.toLowerCase().indexOf(text.toLowerCase()) !== -1;
        //in ES7 / ES2016 we have .includes(), new method better from .indexOf() in many cases:
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        // .includes() search also for NaN and undefined (indexOf don't)

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

// TIMESTAMPS (in milliseconds, 1s = 1000 ms )
// start 0 = 01.01.1970 (unix epoch)
// 1000 = 1000 milliseconds = 1s after 01.01.1970
// -60000 = 60s = 1min before 01.01.1970