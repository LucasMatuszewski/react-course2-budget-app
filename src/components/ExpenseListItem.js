import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
/* // REFACTOR: move "remove" button to EditExpense.js (connect & removeEx no more needed here)
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses'; */

/* // MY VERSION:
export default (expense) => (
    <li>
        {expense.description}
        , {(expense.amount / 100).toLocaleString("en-US", {style:"currency", currency:"USD"})} 
        , added: {(new Date(expense.createdAt)).toLocaleDateString()}<br />
        note: {expense.note}
    </li>
); */

// UDEMY VERSION:
const ExpenseListItem = ({ id, description, amount, createdAt, note}) => (
/* const ExpenseListItem = (props) => ( */
    <div>
        {/* <Link to={'/edit/' + id } title="Edit this expense"> */} {/* MY VERSION */}
        <Link to={`/edit/${id}`} title="Edit this expense">
            <h3>{description}</h3>
        </Link>
        <p>
            {/* // MY VERSION WITHOUT MOMENT() ==> Date().toLocaleDateString()
            {(amount / 100).toLocaleString("en-US", { style: "currency", currency: "USD" })}
            , {(new Date(createdAt)).toLocaleDateString()}<br />
            note: {note} */}
            
            {/* // UDEMY VERSION: */}
            { numeral(amount / 100).format('$0,0.00') 
                // http://numeraljs.com/

                // Udemy's instructor use Module: Numerl.js - but why???
                // toLocaleString and toLocaleDateString works fine. Why use external libraries?
                // It only make a bundle.js bigger...
            }
             - 
            { moment(createdAt).format('MMMM Do, YYYY') // June 30th, 2018
                // https://momentjs.com/docs/#/displaying/format/
            }
        </p>

        {/* REFACTOR: move "remove" button to EditExpense.js */}
        {/* <button onClick={() => { */}
            {/* console.log(props.id); */}
            {/* dispatch(removeExpense({id: id})); */}
            {/* dispatch(removeExpense({ id })); */} {/* ES6 Object definition shorthand for "id: id" */}
        {/* }} >Remove</button> */}
        {/* <button
            onClick={() => {

            }}
        >Edit</button> */}

    </div>
);


// export default connect()(ExpenseListItem);
// Refactor: move "remove" button to EditExpense.js (connect no more needed here)
export default ExpenseListItem;

// https://stackoverflow.com/questions/12196689/how-to-convert-milliseconds-to-a-date-string
// .toDateString()
// .toLocaleDateString()
// https://javascript.info/date

// https://stackoverflow.com/questions/32768494/convert-a-whole-number-amount-of-cents-to-a-readable-dollar-amount-in-javascript/32768579
// .toLocaleString("en-US", {style:"currency", currency:"USD"})

// .map() vs .forEach():
// https://codeburst.io/javascript-map-vs-foreach-f38111822c0f