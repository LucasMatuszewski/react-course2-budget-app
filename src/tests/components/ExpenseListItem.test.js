import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses'; //dummy data for tests

test('should render ExpenseListItem with 1 dummy expense: Rent', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[1]} />);
    expect(wrapper).toMatchSnapshot();
});
