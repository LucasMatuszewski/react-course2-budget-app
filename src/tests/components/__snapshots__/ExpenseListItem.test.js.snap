// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with 1 dummy expense: Rent 1`] = `
<div>
   
  <Link
    replace={false}
    title="Edit this expense"
    to="/edit/2"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    $12,123.41
    -
    December 28th, 1969
  </p>
   
</div>
`;
